daceml/pytorch/module_codegen.py

```diff
@@ -148,51 +185,206 @@ def item_to_cpp_literal(item) -> str:
+    for i, o in enumerate(outputs):
+        grad_name = backward_result.given_grad_names[o]
+        code += f'\nauto {grad_name}_ = grad_outputs[{i}];'
+
+    return code
 
 
 def code_for_backward_function(module: 'daceml.pytorch.DaceModule',
                                forward_sdfg: dace.SDFG,
-                               backward_sdfg: dace.SDFG) -> str:
-    pass
+                               backward_sdfg: dace.SDFG,
+                               backward_result: BackwardResult,
+                               forwarded_arrays: Dict[str, data.Data]) -> str:
+
+    inputs, outputs = get_arglist(module)
+    sdfg_name = forward_sdfg.name
+
+    ret_str = return_type_str(outputs)
+
+    outputs_with_forwarded_outputs = copy.deepcopy(outputs)
+    outputs_with_forwarded_outputs.extend(
+        n for n in forwarded_arrays if n not in inputs and n not in outputs)
+
+    fwd_ptr_init_code, fwd_sdfg_call_arguments, _ = argument_codegen(
+        forward_sdfg, module.dace_model.clean_weights, inputs,
+        outputs_with_forwarded_outputs)
 
+    # inputs are given_grads + forwarded_outputs
+    bwd_inputs = list(
+        backward_result.given_grad_names.values()) + list(forwarded_arrays)
 
-def code_for_module(module: 'daceml.pytorch.DaceModule') -> str:
+    # outputs are required grads
+    bwd_outputs = list(backward_result.required_grad_names.values())
+
+    bwd_ptr_init_code, bwd_sdfg_call_arguments, _ = argument_codegen(
+        backward_sdfg,
+        module.dace_model.clean_weights,
+        bwd_inputs,
+        bwd_outputs,
+        guard_contiguous=list(backward_result.given_grad_names.values()))
+
+    # saved inputs/outputs
+    saved_io_for_backward = [
+        n for n in forwarded_arrays if n in inputs or n in outputs
+    ]
+    other_saved_for_backward = [
+        n for n in forwarded_arrays if n not in inputs and n not in outputs
+    ]
+    return f"""
+{get_header(forward_sdfg, backward_sdfg, inputs, outputs, module.use_cuda)}
+class {sdfg_name}Function : public torch::autograd::Function<{sdfg_name}Function> {{
+    public:
+        static
+            {ret_str}
+            forward(
+            AutogradContext *ctx,
+            int64_t fwd_handle_ptr, int64_t bwd_handle_ptr, {", ".join(f"const Tensor& {name}_" for name in inputs)}) {{
+
+            at::AutoNonVariableTypeMode g;
+
+            // initialize outputs
+            {initialize_outputs_code(module, outputs_with_forwarded_outputs)}
+
+            {fwd_ptr_init_code}
+
+            // get SDFG state handle
+            {forward_sdfg.name}Handle_t handle = reinterpret_cast<{forward_sdfg.name}Handle_t>(fwd_handle_ptr);
+
+
+            // call SDFG
+            __program_{forward_sdfg.name}(handle, {fwd_sdfg_call_arguments});
+
+            // save inputs/outputs for backward
+            {
+                f"ctx->save_for_backward({{{', '.join(f'{n}' for n in saved_io_for_backward)}}});" 
+                if saved_io_for_backward else ""
```

Comment:
is this an f-string in an f-string?

Reply:
It's actually an f-string within an f-string within an f-string